File: godfather/factions/neutral/serial_killer.py
from godfather.game import Game
from godfather.factions.base import Faction

OPPOSING_FACTIONS = ['neutral.arsonist', 'mafia', 'town']


class SerialKillerNeutral(Faction):
    name = 'Serial Killer'
    id = 'neutral.serialkiller'
    win_con = 'Be the last player standing.'

    @property
    def category_name(self):
        return 'Neutral'

    def has_won(self, game: Game):
        alive_sks = len(game.players.filter(
            faction='neutral.serialkiller', is_alive=True))
        alive_opposing = len(
            [*filter(lambda pl: pl.role.faction.id in OPPOSING_FACTIONS and pl.is_alive, game.players)])
        return alive_sks > 0 and alive_opposing == 0

File: src/lib/mafia/factions/neutral/SerialKiller.ts
import Faction from '@mafia/structures/Faction';
import type Game from '@mafia/structures/Game';

const WINS_WITH = ['Juggernaut', 'Survivor'];

export default class SerialKillerFaction extends Faction {
	public name = 'Serial Killer';
	public winCondition = 'game/factions:nkWinCondition';

	public hasWon(game: Game) {
		const { players } = game;

		const aliveSerialKillers = players.filter((player) => player.isAlive && player.role.faction.name === 'Serial Killer');
		const aliveOpposing = players.filter((player) => player.isAlive && !WINS_WITH.includes(player.role!.faction.name));

		return aliveSerialKillers.length > 0 && aliveOpposing.length === 0;
	}
}
